File: app/src/main/java/com/valyakinaleksey/roleplayingsystem/modules/gamescreen/submodules/masterlogscreen/domain/interactor/MasterLogUseCase.java
package com.valyakinaleksey.roleplayingsystem.modules.gamescreen.submodules.masterlogscreen.domain.interactor;

import com.ezhome.rxfirebase2.database.RxFirebaseDatabase;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.valyakinaleksey.roleplayingsystem.modules.gamescreen.submodules.masterlogscreen.domain.model.MasterLogMessage;
import com.valyakinaleksey.roleplayingsystem.modules.gameslist.domain.model.GameModel;
import com.valyakinaleksey.roleplayingsystem.utils.FireBaseUtils;

import java.util.HashMap;
import org.joda.time.DateTime;
import rx.Observable;

import static com.valyakinaleksey.roleplayingsystem.utils.FireBaseUtils.ID;
import static com.valyakinaleksey.roleplayingsystem.utils.FireBaseUtils.TEMP_DATE_CREATE;

public class MasterLogUseCase implements MasterLogInteractor {

  @Override public Observable<MasterLogMessage> sendMessage(GameModel gameModel,
      MasterLogMessage masterLogMessage) {
    DatabaseReference reference = getReference(gameModel);
    masterLogMessage.setDateCreate(DateTime.now().getMillis());
    return RxFirebaseDatabase.getInstance()
        .observeSetValuePush(reference, masterLogMessage)
        .map(s -> {
          DatabaseReference child = reference.child(s);
          HashMap<String, Object> map = new HashMap<>();
          map.put(TEMP_DATE_CREATE, null);
          map.put(ID, s);
          child.updateChildren(map);
          masterLogMessage.setId(s);
          return masterLogMessage;
        });
  }

  private DatabaseReference getReference(GameModel gameModel) {
    return FirebaseDatabase.getInstance()
        .getReference()
        .child(FireBaseUtils.GAME_LOG)
        .child(gameModel.getId());
  }
}
      

File: app/src/main/java/com/valyakinaleksey/roleplayingsystem/modules/gamescreen/submodules/masterlogscreen/domain/interactor/MasterLogUseCase.kt
package com.valyakinaleksey.roleplayingsystem.modules.gamescreen.submodules.masterlogscreen.domain.interactor

import com.valyakinaleksey.roleplayingsystem.data.repository.game.log.LogRepository
import com.valyakinaleksey.roleplayingsystem.modules.gamescreen.domain.model.GameModel
import com.valyakinaleksey.roleplayingsystem.modules.gamescreen.submodules.masterlogscreen.domain.model.MasterLogMessage
import com.valyakinaleksey.roleplayingsystem.modules.gamescreen.submodules.masterlogscreen.domain.model.MasterLogMessageViewModel
import eu.davidea.flexibleadapter.items.IFlexible
import rx.Observable

class MasterLogUseCase(private val logRepository: LogRepository) : MasterLogInteractor {

    override fun sendMessage(gameModel: GameModel, masterLogMessage: MasterLogMessage): Observable<MasterLogMessage> {
        return logRepository.sendMessage(gameModel.id, masterLogMessage)
    }

    override fun observeMessages(gameId: String): Observable<List<IFlexible<*>>> {
        return logRepository.observeData(gameId)
                .map { messages ->
                    val result = ArrayList<IFlexible<*>>()
                    messages.values.reversed().forEach { message ->
                        result.add(MasterLogMessageViewModel(message))
                    }

                    return@map result
                }
    }

}
      